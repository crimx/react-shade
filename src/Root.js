// @flow

import React, { Component, type Node } from "react";
import { createPortal } from "react-dom";
import Context from "./Context";
import retarget from "./internal/retarget";

type Props = {
  children?: Node,
  tag: string
};

type State = {
  shadowRoot?: window.Node
};

export class Root extends Component<Props, State> {
  static defaultProps = {
    tag: "div"
  };
  state = {};
  attachShadow: Function = (e: HTMLElement): void => {
    if (e) {
      const shadowRoot = e.attachShadow({ mode: "open" });
      retarget(shadowRoot);
      this.setState({ shadowRoot });
    }
  };
  render() {
    const { attachShadow, props, state } = this;
    const { tag: Tag, ...rest } = this.props;
    return (
      <Context.Provider value={state.shadowRoot}>
        <Tag {...rest} ref={attachShadow}>
          {state.shadowRoot
            ? createPortal(props.children, state.shadowRoot)
            : null}
        </Tag>
      </Context.Provider>
    );
  }
}
