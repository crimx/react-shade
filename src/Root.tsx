import * as React from "react";
import { createPortal } from "react-dom";
import Context from "./Context";
import retarget from "./internal/retarget";

// This is so that we can have a dynamic tag name and still pass ref.
declare module "react" {
  namespace JSX {
    interface IntrinsicAttributes {
      ref?: any;
    }
  }
}

type Props = {
  children?: React.ReactChildren;
  tag: string;
};

type State = {
  shadowRoot?: Node;
};

export class Root extends React.Component<Props, State> {
  static defaultProps = {
    tag: "div"
  };
  state = {
    shadowRoot: null
  };
  attachShadow = (e: HTMLElement): void => {
    if (e) {
      const shadowRoot = e.attachShadow({ mode: "open" });
      retarget(shadowRoot);
      this.setState({ shadowRoot });
    }
  };
  render() {
    const { attachShadow, props, state } = this;
    const { tag: Tag, ...rest } = this.props;
    return (
      <Context.Provider value={state.shadowRoot}>
        <Tag {...rest} ref={attachShadow}>
          {state.shadowRoot
            ? createPortal(props.children, state.shadowRoot)
            : null}
        </Tag>
      </Context.Provider>
    );
  }
}
