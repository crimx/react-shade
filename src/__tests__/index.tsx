import * as React from "react";
import { render as reactDomRender } from "react-dom";
import Root, { Slot } from "..";

document.body.innerHTML = "<div></div>";

function render(node): HTMLElement {
  const { firstChild } = document.body;
  reactDomRender(node, firstChild);

  // The wrapper creates a wrapping node because ReactDOM's createPortal()
  // function can't render to a shadow root, so we must reach in and return
  // the wrapper it generates.
  return firstChild.firstChild;
}

test("creates a shadow root", () => {
  const root = render(<Root />);
  expect(root.shadowRoot).toBeDefined();
});

test("renders text into the shadow root", () => {
  const root = render(<Root>test</Root>);
  expect(root.shadowRoot.innerHTML).toMatchSnapshot();
});

test("renders elements into the shadow root", () => {
  const root = render(
    <Root>
      <div>test</div>
    </Root>
  );
  expect(root.shadowRoot.innerHTML).toMatchSnapshot();
});

test("rendered inside of another node", () => {
  const root = render(
    <section>
      <Root>
        <div>test</div>
      </Root>
    </section>
  );
  expect(root.outerHTML).toMatchSnapshot();
  expect(root.firstChild.shadowRoot.innerHTML).toMatchSnapshot();
});

test("props - tag", () => {
  const root = render(<Root tag="span">test</Root>);
  expect(root.outerHTML).toMatchSnapshot();
  expect(root.shadowRoot.innerHTML).toMatchSnapshot();
});

test("Slot - renders a slot", () => {
  const root = render(
    <Root>
      <Slot />
    </Root>
  );
  expect(root.outerHTML).toMatchSnapshot();
  expect(root.shadowRoot.innerHTML).toMatchSnapshot();
});

test("Slot - renders multiple slots", () => {
  const root = render(
    <Root>
      <Slot />
      <Slot />
      <Slot />
    </Root>
  );
  expect(root.outerHTML).toMatchSnapshot();
  expect(root.shadowRoot.innerHTML).toMatchSnapshot();
});

test("Slot - slots content", () => {
  const root = render(
    <Root>
      <Slot>content 1</Slot>
      <Slot>
        <span>content 2</span>
      </Slot>
      <Slot>
        <span>content 3</span>
      </Slot>
    </Root>
  );

  // Sanity check for the host.
  expect(root.outerHTML).toMatchSnapshot();

  // To look at the slots.
  expect(root.shadowRoot.innerHTML).toMatchSnapshot();

  // To look at where the light DOM is rendered.
  expect(root.parentNode.outerHTML).toMatchSnapshot();
});
