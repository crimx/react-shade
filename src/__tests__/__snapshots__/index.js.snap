// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slot - renders a slot 1`] = `"<div></div>"`;

exports[`Slot - renders a slot 2`] = `"<slot name=\\"slot-1\\"></slot>"`;

exports[`Slot - renders multiple slots 1`] = `"<div></div>"`;

exports[`Slot - renders multiple slots 2`] = `"<slot name=\\"slot-1\\"></slot><slot name=\\"slot-2\\"></slot><slot name=\\"slot-3\\"></slot>"`;

exports[`Slot - slots content 1`] = `"<div><span slot=\\"slot-1\\">content 1</span><span slot=\\"slot-2\\">content 2</span><span slot=\\"slot-3\\">content 3</span></div>"`;

exports[`Slot - slots content 2`] = `"<slot name=\\"slot-1\\"></slot><slot name=\\"slot-2\\"></slot><slot name=\\"slot-3\\"></slot>"`;

exports[`Slot - slots content 3`] = `"<div><div><span slot=\\"slot-1\\">content 1</span><span slot=\\"slot-2\\">content 2</span><span slot=\\"slot-3\\">content 3</span></div></div>"`;

exports[`props - tag 1`] = `"<span></span>"`;

exports[`props - tag 2`] = `"test"`;

exports[`rendered inside of another node 1`] = `"<section><div></div></section>"`;

exports[`rendered inside of another node 2`] = `"<div>test</div>"`;

exports[`renders elements into the shadow root 1`] = `"<div>test</div>"`;

exports[`renders text into the shadow root 1`] = `"test"`;
